File: src/backend/llvm/emit_ll.py
"""CIR -> textual LLVM IR.

Module M5.  Owner: Member 3.
Status: IMPLEMENTED for every CIR opcode in docs/cir-spec.md section 3.

Why textual IR and not the LLVM C API
-------------------------------------
The artefact under review is the translation, not the bindings.  Text is
diffable, is what the report's figures show, and is checked by ``llvm-as``
in CI -- which is a real verifier, not a smoke test.  It also keeps the
project dependency-free, matching the choice already made in requirements.txt.

The shape of the pass
---------------------
CIR's CFG maps one-to-one onto LLVM's: both allow a branch to any label in
the function, so no restructuring is needed and this back end is close to a
printer.  That is the contrast the report draws against the WebAssembly back
end (M6b), which has to rebuild the control flow as nested regions.

What stops it being *only* a printer is guards.py.  Four CIR operations mean
something LLVM leaves undefined, so this emitter splits the current block and
branches to a trap block instead of emitting the bare LLVM instruction.  A
block split is why the emitter works on a flat line buffer with its own label
counter rather than emitting one LLVM block per CIR block.
"""

from __future__ import annotations

from typing import Dict, List, Optional

from ...cir.ir import (ConstFloat, ConstInt, Function, GlobalRef, INT_WIDTH,
                       Instr, Module, Reg, Ty, Value, is_cmp, wrap_int)
from . import guards
from .typemap import llvm_ty

# CIR opcode -> LLVM binary instruction, for the cases that are a rename.
_BINOP = {
    "add": "add", "sub": "sub", "mul": "mul",
    "sdiv": "sdiv", "udiv": "udiv", "srem": "srem", "urem": "urem",
    "fadd": "fadd", "fsub": "fsub", "fmul": "fmul", "fdiv": "fdiv",
    "and": "and", "or": "or", "xor": "xor",
    "shl": "shl", "ashr": "ashr", "lshr": "lshr",
}
_CONV = {"sext": "sext", "zext": "zext", "trunc": "trunc",
         "sitofp": "sitofp", "fptosi": "fptosi"}
_DIV_OPS = ("sdiv", "udiv", "srem", "urem")
_SHIFT_OPS = ("shl", "ashr", "lshr")

_FMT = {"print.i32": ("@.fmt.i32", "i32"), "print.f64": ("@.fmt.f64", "double")}


class EmitError(Exception):
    """CIR this back end cannot lower.  Always a bug in the input, not here."""


def fmt_float(value: float) -> str:
    """An LLVM double literal.

    LLVM's lexer only treats a number as floating point once it has seen a
    '.', so an exponent form such as ``1e-05`` would lex as the integer 1.
    Infinities and NaN have no decimal spelling at all and use the raw 64-bit
    hexadecimal form.
    """
    if value != value or value in (float("inf"), float("-inf")):
        import struct
        bits = struct.unpack("<Q", struct.pack("<d", value))[0]
        return f"0x{bits:016X}"
    text = repr(float(value))
    if "." not in text:
        text = text.replace("e", ".0e") if "e" in text else text + ".0"
    return text


class Emitter:
    """Emits one function into a flat line buffer.

    Guards split basic blocks, so the buffer holds already-formatted LLVM
    lines and label lines rather than a list of blocks.  ``tmp`` and ``label``
    hand out names under the ``g.`` prefix, which the builder never generates
    -- its registers are ``%tN`` or a source name with an ``.addr`` suffix.
    """

    def __init__(self, sigs: Dict[str, tuple], needs: set) -> None:
        self.sigs = sigs
        self.needs = needs
        self.out: List[str] = []
        self.n = 0
        self.extents: Dict[str, int] = {}

    # -- output ------------------------------------------------------------
    def line(self, text: str) -> None:
        self.out.append(f"  {text}")

    def label_line(self, name: str) -> None:
        self.out.append(f"{name}:")

    def tmp(self) -> str:
        self.n += 1
        return f"%g.{self.n}"

    def label(self, base: str) -> str:
        self.n += 1
        return f"{base}.{self.n}"

    def trap_branch(self, cond: str, kind: str) -> None:
        """Branch to a fresh trap block when ``cond`` holds, and carry on.

        The continuation label is opened immediately, so the caller can keep
        emitting into what is now a new LLVM basic block.
        """
        self.needs.add("trap")
        trap_l = self.label(f"trap.{kind}")
        cont_l = self.label(f"cont.{kind}")
        self.line(f"br i1 {cond}, label %{trap_l}, label %{cont_l}")
        self.label_line(trap_l)
        self.line("call void @llvm.trap()")
        self.line("unreachable")
        self.label_line(cont_l)

    # -- values ------------------------------------------------------------
    def val(self, value: Value, ty: Ty) -> str:
        if isinstance(value, Reg):
            return f"%{value.name}"
        if isinstance(value, GlobalRef):
            return f"@{value.name}"
        if isinstance(value, ConstFloat):
            return fmt_float(value.value)
        if isinstance(value, ConstInt):
            if ty is Ty.I1:
                return "true" if value.value & 1 else "false"
            if ty is Ty.F64:
                return fmt_float(float(value.value))
            return str(wrap_int(value.value, ty) if ty in INT_WIDTH
                       else value.value)
        raise EmitError(f"cannot emit operand {value!r}")

    def ty_of(self, value: Value, default: Ty) -> Ty:
        if isinstance(value, GlobalRef):
            return Ty.PTR
        return getattr(value, "ty", default)

    # -- functions ---------------------------------------------------------
    def function(self, fn: Function) -> str:
        self.out = []
        self.n = 0
        self.extents = {
            i.dest.name: i.args[0].value
            for b in fn.blocks for i in b.instrs
            if i.op == "alloca" and i.dest is not None and i.args
            and isinstance(i.args[0], ConstInt)
        }

        params = ", ".join(f"{llvm_ty(p.ty)} %{p.name}" for p in fn.params)
        head = f"define {llvm_ty(fn.ret_ty)} @{fn.name}({params}) {{"

        for block in fn.blocks:
            self.label_line(block.label)
            for instr in block.instrs:
                self.instr(instr)
        return "\n".join([head] + self.out + ["}"])

    # -- instructions ------------------------------------------------------
    def instr(self, ins: Instr) -> None:
        op = ins.op
        ty = ins.ty
        lty = llvm_ty(ty)
        dest = f"%{ins.dest.name}" if ins.dest is not None else None
        args = [self.val(a, ty) for a in ins.args]

        if op == "br":
            self.line(f"br label %{ins.labels[0]}")
            return
        if op == "br.cond":
            self.line(f"br i1 {args[0]}, label %{ins.labels[0]}, "
                      f"label %{ins.labels[1]}")
            return
        if op == "ret":
            self.line("ret void" if not ins.args else f"ret {lty} {args[0]}")
            return
        if op == "trap":
            # Not a CIR terminator, so no 'unreachable' here: the block's own
            # terminator still follows.  llvm.trap is noreturn, so control
            # never actually reaches it.
            self.needs.add("trap")
            self.line("call void @llvm.trap()")
            return

        if op in _FMT:
            self.needs.add("printf")
            symbol, arg_ty = _FMT[op]
            self.line(f"call i32 (ptr, ...) @printf(ptr {symbol}, "
                      f"{arg_ty} {args[0]})")
            return

        if op == "call":
            self.call(ins, dest, lty)
            return

        if op == "alloca":
            count = f", i32 {args[0]}" if ins.args else ""
            self.line(f"{dest} = alloca {lty}{count}")
            return
        if op == "load":
            self.line(f"{dest} = load {lty}, ptr {args[0]}")
            return
        if op == "store":
            self.line(f"store {lty} {args[0]}, ptr {args[1]}")
            return
        if op == "gep":
            base = ins.args[0]
            extent = self.extents.get(base.name) if isinstance(base, Reg) else None
            index_text = self.val(ins.args[1], Ty.I32)
            if extent is not None:
                guards.guard_bounds(self, extent, ins.args[1], index_text)
            self.line(f"{dest} = getelementptr {lty}, ptr {args[0]}, "
                      f"i32 {index_text}")
            return

        if is_cmp(op):
            kind, _, pred = op.partition(".")
            # CIR prints the *operand* type on a comparison; the result is i1.
            self.line(f"{dest} = {kind} {pred} {lty} {args[0]}, {args[1]}")
            return

        if op in _CONV:
            src = llvm_ty(self.ty_of(ins.args[0], Ty.I32))
            if op == "fptosi":
                guards.guard_fptosi(self, ty, args[0])
            self.line(f"{dest} = {_CONV[op]} {src} {args[0]} to {lty}")
            return

        if op == "neg":
            if ty is Ty.F64:
                self.line(f"{dest} = fneg double {args[0]}")
            else:
                # LLVM has no integer negate; CIR's 'neg' becomes 0 - x, which
                # is what the builder emits directly for integers anyway.
                self.line(f"{dest} = sub {lty} 0, {args[0]}")
            return
        if op == "not":
            ones = "true" if ty is Ty.I1 else "-1"
            self.line(f"{dest} = xor {lty} {args[0]}, {ones}")
            return

        if op in _BINOP:
            rhs = args[1]
            if op in _DIV_OPS:
                guards.guard_div(self, op, ty, args[0], ins.args[1], rhs)
            elif op in _SHIFT_OPS:
                rhs = guards.mask_shift_count(self, ty, ins.args[1], rhs)
            # Deliberately no 'nsw'/'nuw': CIR defines signed overflow as
            # wraparound (docs/divergence.md row 4), and those flags would
            # make it poison instead.
            self.line(f"{dest} = {_BINOP[op]} {lty} {args[0]}, {rhs}")
            return

        raise EmitError(f"no LLVM lowering for CIR opcode {op!r}")

    def call(self, ins: Instr, dest: Optional[str], lty: str) -> None:
        sig = self.sigs.get(ins.callee)
        param_tys = sig[1] if sig else [self.ty_of(a, Ty.I32) for a in ins.args]
        if len(param_tys) != len(ins.args):
            raise EmitError(
                f"@{ins.callee} takes {len(param_tys)} argument(s), "
                f"{len(ins.args)} given")
        args = ", ".join(f"{llvm_ty(t)} {self.val(a, t)}"
                         for a, t in zip(ins.args, param_tys))
        call = f"call {lty} @{ins.callee}({args})"
        self.line(call if dest is None else f"{dest} = {call}")


# --------------------------------------------------------------------------
# Module
# --------------------------------------------------------------------------
_RUNTIME = {
    "printf": [
        'declare i32 @printf(ptr, ...)',
        '@.fmt.i32 = private unnamed_addr constant [4 x i8] c"%d\\0A\\00"',
        '@.fmt.f64 = private unnamed_addr constant [4 x i8] c"%f\\0A\\00"',
    ],
    "trap": ['declare void @llvm.trap()'],
}


def emit(module: Module) -> str:
    """Emit textual LLVM IR for a CIR module.

    The runtime declarations at the end are emitted only when something used
    them, so a module that neither prints nor traps produces no unused
    declarations and the golden files stay small.
    """
    sigs = {f.name: (f.ret_ty, [p.ty for p in f.params]) for f in module.functions}
    needs: set = set()
    emitter = Emitter(sigs, needs)

    bodies = [emitter.function(fn) for fn in module.functions]

    # Shared verbatim with the C++ back end in src/backend/llvm/EmitLL.cpp so
    # that docs/examples/abs.gen.ll is a golden file for both.
    parts = [f"; CIR module {module.name!r} lowered by the MTIR LLVM back end "
             f"(M5)"]
    for g in module.globals:
        gty = llvm_ty(g.ty)
        init = emitter.val(g.init, g.ty) if g.init is not None else "zeroinitializer"
        if g.array_len is not None:
            parts.append(f"@{g.name} = global [{g.array_len} x {gty}] "
                         f"zeroinitializer")
        else:
            parts.append(f"@{g.name} = global {gty} {init}")
    if module.globals:
        parts.append("")

    parts += bodies

    runtime = [line for key in ("printf", "trap") if key in needs
               for line in _RUNTIME[key]]
    if runtime:
        parts.append("")
        parts += runtime

    return "\n".join(parts) + "\n"

File: src/backend/llvm/EmitLL.cpp
#include "mtir/backend/llvm/EmitLL.h"

#include <cmath>
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <set>
#include <unordered_map>
#include <utility>
#include <vector>

#include "mtir/backend/llvm/Guards.h"
#include "mtir/backend/llvm/TypeMap.h"
#include "mtir/cir/Arith.h"
#include "mtir/cir/Printer.h"

namespace mtir::backend::llvm {
namespace {

using namespace mtir::cir;
using support::Diagnostics;
using support::Location;

std::string ty(Ty t) { return std::string(llvmType(t)); }

/// The LLVM instruction for a CIR opcode that is a straight rename.
const char *binaryOpName(Opcode op) {
  switch (op) {
  case Opcode::Add: return "add";
  case Opcode::Sub: return "sub";
  case Opcode::Mul: return "mul";
  case Opcode::SDiv: return "sdiv";
  case Opcode::UDiv: return "udiv";
  case Opcode::SRem: return "srem";
  case Opcode::URem: return "urem";
  case Opcode::FAdd: return "fadd";
  case Opcode::FSub: return "fsub";
  case Opcode::FMul: return "fmul";
  case Opcode::FDiv: return "fdiv";
  case Opcode::And: return "and";
  case Opcode::Or: return "or";
  case Opcode::Xor: return "xor";
  case Opcode::Shl: return "shl";
  case Opcode::AShr: return "ashr";
  case Opcode::LShr: return "lshr";
  default: return nullptr;
  }
}

const char *conversionOpName(Opcode op) {
  switch (op) {
  case Opcode::SExt: return "sext";
  case Opcode::ZExt: return "zext";
  case Opcode::Trunc: return "trunc";
  case Opcode::SIToFP: return "sitofp";
  case Opcode::FPToSI: return "fptosi";
  default: return nullptr;
  }
}

bool isDivision(Opcode op) {
  return op == Opcode::SDiv || op == Opcode::UDiv || op == Opcode::SRem ||
         op == Opcode::URem;
}

bool isShift(Opcode op) {
  return op == Opcode::Shl || op == Opcode::AShr || op == Opcode::LShr;
}

bool isIntCompare(Opcode op) {
  return op >= Opcode::ICmpEq && op <= Opcode::ICmpUge;
}

/// A double in LLVM's spelling.
///
/// This is *not* cir::printDouble.  The two formats agree on finite values but
/// not on the rest: CIR text spells them `inf`, `-inf` and `nan`, which its own
/// parser reads back, whereas LLVM has no such literal and requires the raw
/// 64-bit pattern.  Emitting `double inf` produces a file llvm-as rejects, and
/// constant folding really can reach it -- `fmul f64 1.0e308, 10.0` overflows.
std::string llvmDouble(double value) {
  if (std::isnan(value) || std::isinf(value)) {
    std::uint64_t bits = 0;
    static_assert(sizeof(bits) == sizeof(value), "double must be 64 bits");
    std::memcpy(&bits, &value, sizeof(bits));
    char buffer[32];
    std::snprintf(buffer, sizeof(buffer), "0x%016llX",
                  static_cast<unsigned long long>(bits));
    return buffer;
  }
  return printDouble(value);
}

/// One operand in LLVM spelling.  `context` is the type LLVM will read the
/// operand as, which is what turns an i1 constant into true/false and wraps
/// an out-of-range integer literal into its type (docs/divergence.md row 4).
std::string llvmConstant(const Value &v, Ty context) {
  struct Visitor {
    Ty context;
    std::string operator()(const Reg &r) const { return "%" + r.name; }
    std::string operator()(const GlobalRef &g) const { return "@" + g.name; }
    std::string operator()(const ConstFloat &c) const { return llvmDouble(c.value); }
    std::string operator()(const ConstInt &c) const {
      if (context == Ty::I1)
        return (c.value & 1) != 0 ? "true" : "false";
      if (context == Ty::F64)
        return llvmDouble(static_cast<double>(c.value));
      if (isInteger(context))
        return std::to_string(wrapInt(c.value, context));
      return std::to_string(c.value);
    }
  };
  return std::visit(Visitor{context}, v);
}

// --------------------------------------------------------------------------
// The emitter
// --------------------------------------------------------------------------
class FunctionEmitter final : public Emitter {
public:
  FunctionEmitter(const Module &module, std::set<std::string> &needs,
                  Diagnostics &diags)
      : module_(module), needs_(needs), diags_(diags) {}

  std::string emit(const Function &fn);

  // -- Emitter interface -------------------------------------------------
  std::string tmp() override { return "%g." + std::to_string(++counter_); }

  void line(std::string_view text) override {
    out_.push_back("  " + std::string(text));
  }

  void trapBranch(std::string_view cond, std::string_view kind) override {
    needs_.insert("trap");
    const std::string suffix = std::to_string(++counter_);
    const std::string trapLabel = "trap." + std::string(kind) + "." + suffix;
    const std::string contLabel = "cont." + std::string(kind) + "." + suffix;
    line("br i1 " + std::string(cond) + ", label %" + trapLabel + ", label %" +
         contLabel);
    label(trapLabel);
    line("call void @llvm.trap()");
    line("unreachable");
    label(contLabel);
  }

private:
  void label(const std::string &name) { out_.push_back(name + ":"); }

  void error(const Function &fn, const BasicBlock &block, std::string message) {
    Location loc;
    loc.function = fn.name();
    loc.block = block.label();
    diags_.push_back(support::error("LLVM01", std::move(message), loc));
  }

  std::string value(const Value &v, Ty context) const;
  void instruction(const Function &fn, const BasicBlock &block,
                   const Instruction &instr);
  void call(const Function &fn, const BasicBlock &block, const Instruction &instr);

  const Module &module_;
  std::set<std::string> &needs_;
  Diagnostics &diags_;
  std::vector<std::string> out_;
  std::unordered_map<std::string, std::int64_t> extents_;
  unsigned counter_ = 0;
};

std::string FunctionEmitter::value(const Value &v, Ty context) const {
  return llvmConstant(v, context);
}

std::string FunctionEmitter::emit(const Function &fn) {
  out_.clear();
  counter_ = 0;
  extents_.clear();

  // Array extents recovered from the allocas in this function, so a gep off a
  // sized alloca can be bounds-checked (docs/divergence.md row 7).
  for (const BasicBlock &block : fn.blocks())
    for (const Instruction &instr : block.instructions())
      if (instr.op == Opcode::Alloca && instr.dest.has_value() && !instr.args.empty())
        if (const ConstInt *c = asConstInt(instr.args[0]))
          extents_[instr.dest->name] = c->value;

  std::string params;
  for (std::size_t i = 0; i < fn.params().size(); ++i) {
    if (i != 0)
      params += ", ";
    params += ty(fn.params()[i].ty) + " %" + fn.params()[i].name;
  }
  const std::string head =
      "define " + ty(fn.returnType()) + " @" + fn.name() + "(" + params + ") {";

  for (const BasicBlock &block : fn.blocks()) {
    label(block.label());
    for (const Instruction &instr : block.instructions())
      instruction(fn, block, instr);
  }

  std::string body = head;
  for (const std::string &l : out_) {
    body += "\n";
    body += l;
  }
  body += "\n}";
  return body;
}

void FunctionEmitter::call(const Function &fn, const BasicBlock &block,
                           const Instruction &instr) {
  const Function *callee = module_.function(instr.callee);
  std::vector<Ty> paramTys;
  if (callee != nullptr) {
    for (const Param &p : callee->params())
      paramTys.push_back(p.ty);
  } else {
    for (const Value &a : instr.args)
      paramTys.push_back(typeOf(a));
  }

  if (paramTys.size() != instr.args.size()) {
    error(fn, block,
          "@" + instr.callee + " takes " + std::to_string(paramTys.size()) +
              " argument(s), " + std::to_string(instr.args.size()) + " given");
    return;
  }

  std::string args;
  for (std::size_t i = 0; i < instr.args.size(); ++i) {
    if (i != 0)
      args += ", ";
    args += ty(paramTys[i]) + " " + value(instr.args[i], paramTys[i]);
  }

  const std::string text =
      "call " + ty(instr.ty) + " @" + instr.callee + "(" + args + ")";
  line(instr.dest.has_value() ? "%" + instr.dest->name + " = " + text : text);
}

void FunctionEmitter::instruction(const Function &fn, const BasicBlock &block,
                                  const Instruction &instr) {
  const std::string dest =
      instr.dest.has_value() ? "%" + instr.dest->name : std::string();

  std::vector<std::string> args;
  args.reserve(instr.args.size());
  for (const Value &a : instr.args)
    args.push_back(value(a, instr.ty));

  switch (instr.op) {
  case Opcode::Br:
    line("br label %" + instr.labels[0]);
    return;

  case Opcode::BrCond:
    line("br i1 " + args[0] + ", label %" + instr.labels[0] + ", label %" +
         instr.labels[1]);
    return;

  case Opcode::Ret:
    line(instr.args.empty() ? "ret void" : "ret " + ty(instr.ty) + " " + args[0]);
    return;

  case Opcode::Trap:
    // 'trap' is not a CIR terminator, so no 'unreachable' here: the block's
    // own terminator still follows it.  llvm.trap is noreturn, so control
    // never actually reaches past it -- but emitting a terminator mid-block
    // would produce IR that llvm-as rejects.
    needs_.insert("trap");
    line("call void @llvm.trap()");
    return;

  case Opcode::PrintI32:
    needs_.insert("printf");
    line("call i32 (ptr, ...) @printf(ptr @.fmt.i32, i32 " + args[0] + ")");
    return;

  case Opcode::PrintF64:
    needs_.insert("printf");
    line("call i32 (ptr, ...) @printf(ptr @.fmt.f64, double " + args[0] + ")");
    return;

  case Opcode::Call:
    call(fn, block, instr);
    return;

  case Opcode::Alloca:
    line(dest + " = alloca " + ty(instr.ty) +
         (instr.args.empty() ? "" : ", i32 " + args[0]));
    return;

  case Opcode::Load:
    line(dest + " = load " + ty(instr.ty) + ", ptr " + args[0]);
    return;

  case Opcode::Store:
    line("store " + ty(instr.ty) + " " + args[0] + ", ptr " + args[1]);
    return;

  case Opcode::GEP: {
    const std::string indexText = value(instr.args[1], Ty::I32);
    if (const Reg *base = asRegister(instr.args[0])) {
      const auto it = extents_.find(base->name);
      if (it != extents_.end())
        guardBounds(*this, it->second, instr.args[1], indexText);
    }
    line(dest + " = getelementptr " + ty(instr.ty) + ", ptr " + args[0] + ", i32 " +
         indexText);
    return;
  }

  case Opcode::Neg:
    if (instr.ty == Ty::F64)
      line(dest + " = fneg double " + args[0]);
    else
      // LLVM has no integer negate; CIR's 'neg' becomes 0 - x, which is what
      // the builder emits directly for integers anyway.
      line(dest + " = sub " + ty(instr.ty) + " 0, " + args[0]);
    return;

  case Opcode::Not:
    line(dest + " = xor " + ty(instr.ty) + " " + args[0] + ", " +
         (instr.ty == Ty::I1 ? "true" : "-1"));
    return;

  default:
    break;
  }

  if (isCompare(instr.op)) {
    // CIR prints the *operand* type on a comparison; the result is i1.
    const char *kind = isIntCompare(instr.op) ? "icmp" : "fcmp";
    line(dest + " = " + kind + " " + std::string(info(instr.op).predicate) + " " +
         ty(instr.ty) + " " + args[0] + ", " + args[1]);
    return;
  }

  if (const char *conv = conversionOpName(instr.op)) {
    const Ty sourceTy = typeOf(instr.args[0]);
    // Refuse rather than emit: a conversion with the wrong result type would
    // produce text like `fptosi double %x to double`, which llvm-as rejects,
    // and would drive the fptosi guard's range computation with a width of
    // zero.  Better one diagnostic than an invalid .ll file.
    const bool wantsIntResult =
        instr.op == Opcode::FPToSI || instr.op == Opcode::SExt ||
        instr.op == Opcode::ZExt || instr.op == Opcode::Trunc;
    if (wantsIntResult && !isInteger(instr.ty)) {
      error(fn, block, std::string(mnemonic(instr.op)) +
                           " must produce an integer, found " +
                           std::string(toString(instr.ty)));
      return;
    }
    if (instr.op == Opcode::SIToFP && instr.ty != Ty::F64) {
      error(fn, block, "sitofp must produce f64, found " +
                           std::string(toString(instr.ty)));
      return;
    }
    if (instr.op == Opcode::FPToSI && sourceTy != Ty::F64) {
      error(fn, block, "fptosi takes an f64 operand, found " +
                           std::string(toString(sourceTy)));
      return;
    }
    if (instr.op == Opcode::FPToSI)
      guardFPToSI(*this, instr.ty, args[0]);
    line(dest + " = " + conv + " " + ty(sourceTy) + " " + args[0] + " to " +
         ty(instr.ty));
    return;
  }

  if (const char *bin = binaryOpName(instr.op)) {
    // Same reasoning: a shift or an integer division on a non-integer type is
    // malformed CIR, and emitting it would produce invalid LLVM and a bogus
    // shift mask.
    if ((isShift(instr.op) || isDivision(instr.op)) && !isInteger(instr.ty)) {
      error(fn, block, std::string(mnemonic(instr.op)) +
                           " requires an integer type, found " +
                           std::string(toString(instr.ty)));
      return;
    }
    std::string rhs = args[1];
    if (isDivision(instr.op))
      guardDivision(*this, instr.op, instr.ty, args[0], instr.args[1], rhs);
    else if (isShift(instr.op))
      rhs = maskShiftCount(*this, instr.ty, instr.args[1], rhs);
    // Deliberately no 'nsw'/'nuw': CIR defines signed overflow as wraparound
    // (docs/divergence.md row 4), and those flags would make it poison.
    line(dest + " = " + bin + " " + ty(instr.ty) + " " + args[0] + ", " + rhs);
    return;
  }

  error(fn, block,
        "no LLVM lowering for CIR opcode '" + std::string(mnemonic(instr.op)) + "'");
}

// --------------------------------------------------------------------------
// Runtime declarations, emitted only when something used them
// --------------------------------------------------------------------------
void appendRuntime(const std::set<std::string> &needs, std::vector<std::string> &parts) {
  std::vector<std::string> lines;
  if (needs.count("printf") != 0) {
    lines.emplace_back("declare i32 @printf(ptr, ...)");
    lines.emplace_back(R"(@.fmt.i32 = private unnamed_addr constant [4 x i8] c"%d\0A\00")");
    lines.emplace_back(R"(@.fmt.f64 = private unnamed_addr constant [4 x i8] c"%f\0A\00")");
  }
  if (needs.count("trap") != 0)
    lines.emplace_back("declare void @llvm.trap()");

  if (lines.empty())
    return;
  parts.emplace_back();
  for (std::string &l : lines)
    parts.push_back(std::move(l));
}

} // namespace

EmitResult emitModule(const Module &module) {
  EmitResult result;
  // std::set, not unordered_set: the runtime block is emitted from it, and
  // output must not depend on hash order.
  std::set<std::string> needs;

  FunctionEmitter emitter(module, needs, result.diagnostics);

  std::vector<std::string> bodies;
  bodies.reserve(module.functions().size());
  for (const Function &fn : module.functions())
    bodies.push_back(emitter.emit(fn));

  std::vector<std::string> parts;
  // The header line is shared verbatim with the Python reference
  // implementation, so docs/examples/abs.gen.ll is a golden file for both and
  // any divergence between them shows up as a test failure.
  parts.push_back("; CIR module '" + module.name() +
                  "' lowered by the MTIR LLVM back end (M5)");

  for (const Global &g : module.globals()) {
    const std::string gty = std::string(llvmType(g.ty));
    if (g.arrayLen.has_value()) {
      parts.push_back("@" + g.name + " = global [" + std::to_string(*g.arrayLen) +
                      " x " + gty + "] zeroinitializer");
    } else if (g.init.has_value()) {
      parts.push_back("@" + g.name + " = global " + gty + " " +
                      llvmConstant(*g.init, g.ty));
    } else {
      parts.push_back("@" + g.name + " = global " + gty + " zeroinitializer");
    }
  }
  if (!module.globals().empty())
    parts.emplace_back();

  for (std::string &body : bodies)
    parts.push_back(std::move(body));

  appendRuntime(needs, parts);

  for (std::size_t i = 0; i < parts.size(); ++i) {
    if (i != 0)
      result.ir += "\n";
    result.ir += parts[i];
  }
  result.ir += "\n";
  return result;
}

} // namespace mtir::backend::llvm
